cloud_foundry_api/datadog_checks/cloud_foundry_api/cloud_foundry_api.py

```diff
@@ -130,41 +134,173 @@ def get_oauth_token(self):
-                return events
-
-            # Memorize the last event guid at which we stopped in the previous check run
-            # before we update it during this run
-            last_event_guid = self._last_event_guid
-            for cf_event in payload.get("resources", []):
+                space_name = ""
+                if self._api_version == "v2":
+                    space_name = payload["entity"]["name"]
+                elif self._api_version == "v3":
+                    space_name = payload["name"]
+                self._spaces[space_guid] = space_name
+                return space_name
+            except Exception:
+                self.log.exception("Error getting space name for space %s", space_guid)
+
+    def scroll_api_pages(self, url, params, headers):
+        # type: (str, Dict[str, Any], Dict[str, str]) -> Generator
+        page = 1
+        sc_tags = ["api_url:{}".format(urlparse(self._api_url)[1])] + self._tags
+        raised = False
+        # Wrap inside a try/finally to send the service check even if the generator stops being iterated
+        try:
+            while True:
+                params = copy.deepcopy(params)
+                params["page"] = page
+                self.log.debug("Fetching events page %s", page)
                 try:
-                    dd_event, event_guid, event_ts = parse_event(cf_event, self._api_version, additional_tags)
+                    res = self.http.get(url, params=params, headers=headers)
+                except RequestException:
+                    self.log.exception("Error connecting to the Cloud Controller API: URL %s", url)
+                    self.service_check(API_SERVICE_CHECK_NAME, CloudFoundryApiCheck.CRITICAL, tags=sc_tags)
+                    raised = True
+                    return
+                try:
+                    res.raise_for_status()
+                except HTTPError:
+                    self.log.exception("Error querying Cloud Controller API: URL %s - response %s", url, res.text)
+                    self.service_check(API_SERVICE_CHECK_NAME, CloudFoundryApiCheck.CRITICAL, tags=sc_tags)
+                    raised = True
+                    return
+                try:
+                    payload = res.json()
+                except ValueError:
+                    self.log.exception(
+                        "Error decoding response from the Cloud Controller API: URL %s - response %s", url, res.text
+                    )
+                    self.service_check(API_SERVICE_CHECK_NAME, CloudFoundryApiCheck.CRITICAL, tags=sc_tags)
+                    raised = True
+                    return
+                self.log.trace("Payload received from CC API: %s", payload)
+                yield payload
+                # Fetch next page if any
+                next_url = get_next_url(payload, self._api_version)
+                if not next_url:
+                    break
+                page += 1
+        finally:
```

Comment:
Can't you remove the try/finally and just do the service_check at the end?

Reply:
no because for events for instance, we break before the generator is exhausted, so we never reach the service check line in that case